### DISCUSSION TITLE
Was there a change for sendmail in 14.0.0?

### DISCUSSION THREAD

**Discussion by forumlw06** [arendst/Tasmota#21454] [Q&A]
Hello,

i am using an ESP32 S3 with PSRAM and have been building my own Tasmota with gitpod.

To enable sendmail, i added following to my user_config_overwrite.h
```
//-------------------------------
//-- activate sendmail (see: https://github.com/arendst/Tasmota/issues/14404
// test in console with (ESP8266): =>sendmail [smtp.gmail.com:465:user:password:<from@gmail.com>:<to@gmail.com>:test]Hello world
// test in console with (ESP32  ): =>sendmail [smtp.gmail.com:465:user:password:from@gmail.com:to@gmail.com:test]Hello world
// you must use the sharp clips on esp8266 (not on esp32)
//-------------------------------
#ifndef USE_SENDMAIL
	#define USE_SENDMAIL
#endif

//#define DEBUG_EMAIL_PORT // this is the magic bullet for all those {"Sendmail":"Wrong parameters"}
//#define MAIL_TIMEOUT 2000

#define EMAIL_SERVER "smtp.gmail.com"
#define EMAIL_PORT 465
#define EMAIL_USER "from@gmail.com"
#define EMAIL_PASSWORD "aaaa bbbb cccc dddd" // generated by gmail as application specific password

#ifdef ESP32
	//-- for 13.2.0 and above the new sendmail library needs more stack size on esp32 => increase stacksize 
	#undef SET_ESP32_STACK_SIZE
	#define SET_ESP32_STACK_SIZE (12 * 1024)
	// FROM without <> on ESP32
	#define EMAIL_FROM "from@gmail.com"
#endif

#ifdef ESP8266
	// FROM with <> on ESP8266!!!!
	#define EMAIL_FROM "<from@gmail.com>"
#endif`
```

When i compile this with 13.4.0 (master), sendmail just works fine.
When i use the same user_config_overwrite.h (and the same environment from platformio_tasmota_cenv.ini) in 14.0.0 (master) sendmail does not work any longer ("Wrong Parameters").

As i did not change anything (same user_config_override.h as well as environment in 13.4.0 and 14.0.0) my question is, if there was any modification in 14.0.0 that i should take care of (like the increased stack size change from 13.1 to 13.2 on ESP32).

Thanks in advance!

Rainer

**Comment by arendst** [comment]
Try latest dev branch. There is a fix for timeouts using core 3.

**Reply by forumlw06** [reply]
Thanks,

i rebuild 14.0.0.1 with the exact same entries in user_config_overwrite.h as in 13.4.0.
When i try to send a mail the ESP32 S3 restarts.

This is what is see after reboot (see especially INFO3 section). I did not yet enable the "#define MAIL_TIMEOUT 2000" to get more information. Can do that if it helps to debug what is going on.

```
15:49:09.836 RSL: INFO1 = {"Info1":{"Module":"ESP32S3","Version":"14.0.0.1(ESP32S3)","FallbackTopic":"cmnd/DVES_209114_fb/","GroupTopic":"cmnd/tasmotas/"}}
15:49:09.848 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-209114-4372","IPAddress":"192.168.178.66","IP6Global":"","IP6Local":"fe80::deda:cff:fe20:9114%st1"}}
15:49:09.860 RSL: INFO3 = {"Info3":{"RestartReason":{"Exception":1070552928,"Reason":"Unknown","EPC":"4037bbde","EXCVADDR":"00000000","CallChain":["4037bbdb","4208adfd","42105133"]},"BootCount":14}}
```

**Reply by arendst** [reply]
Reason unknown looks like one of our newly introduced ESP32 watchdogs....

Pls try again with #undef USE_ESP32_WDT

**Reply by forumlw06** [reply]
no more restart with that #undef, but sendmail still reports "wrong parameters"

**Reply by forumlw06** [reply]
BTW: i am also using "#define USE_SCRIPT", not sure if that may interfere

**Reply by arendst** [reply]
I'm trying to pinpoint the issue. For me it seems to fail on invalid application-specific password.

Pls enable mail logging and provide the output here. To do that enable ``#define DEBUG_EMAIL_PORT``.

**Reply by arendst** [reply]
Yep. The needed Application-specific pasword is the culprit.

I made a Google App-Password in my google account page and used it in the sendmail command like:

sendmail [smtp.gmail.com:465:user:app-password-no-spaces:from@gmail.com:to@gmail.com:test]Hello world

**Reply by arendst** [reply]
I also see the reason for your initial watchdog restart; it can take several seconds before the initial google smtp server responds.

I will change the wait loop to feed the watchdog.

**Reply by forumlw06** [reply]
Hi Theo,

thanks for your testing/debugging. Really would be astonished, if the app-password would be the culprit. I am using the same password on a 13.4.0 device which updates me if e.g. my garage door opens (this works fine and reliable on 13.4.0).

I build latest development branch with #define DEBUG_EMAIL_PORT. This is what i get. Looks like it is unable to connect to smtp.gmail.com? Any chance to get even more information why?

```

19:04:28.435 SCR: nv=1, tv=0, vns=6, vmem=35, smem=16384, gmem=496, tmem=16921 
19:04:28.436 HTP: Werkzeuge
19:04:28.679 -------------------- sende email
19:04:28.680 SCR: performs "sendmail [smtp.gmail.com:465:from@gmail.com:app-password-no-spaces:from@gmail.com:to@web.de:test]Hello world"
19:04:28.691 SRC: Rule
19:04:28.691 CMD: Grp 0, Cmd 'SENDMAIL', Idx 1, Len 109, Pld -99, Data '[smtp.gmail.com:465:from@gmail.com:app-password-no-spaces:from@gmail.com:to@web.de:test]Hello world'
19:04:28.692 MAI: Size 96
19:04:28.693 MAI: smtp.gmail.com, 465, from@gmail.com, app-password-no-spaces
19:04:28.704 MAI: from@gmail.com, to@web.de, test, Hello world
19:04:28.705 MAI: #### Connecting to SMTP server...
19:04:46.796 MAI: #### Error, unable to connect to server
19:04:46.798 RSL: RESULT = {"Sendmail":"Wrong parameters"}
19:04:46.812 HTP: Konsole
19:04:56.623 WIF: Prüfe Verbindung…
```

**Reply by forumlw06** [reply]
BTW: i retested with a second gmail account and that app password from which i know this works. Same error "unable to connect to server"

**Comment by arendst** [comment]
So there are at least two issues:

- Watchdogs in mail library
- Mail parameter change

I hope someone can chime in and have a look at the mail lib.

**Reply by forumlw06** [reply]
I rebuild also with "#define MAIL_TIMEOUT 2000" to maybe get more information, but i do not get additional information

I was using a small script to send mail every minute
```
>D
t:timer=0

>S
if (timer==0) {
	print -------------------- sende email
	=>sendmail [*:*:*:*:*:to@web.de:TEST] Bodytest
	timer=60
}
```

with that additional define console log looks like this:

```
16:44:49.666 RSL: config = {"ip":"192.168.178.66","dn":"Tasmota","fn":["Tasmota",null,null,null,null,null,null,null],"hn":"tasmota-209114-4372","mac":"DCDA0C209114","md":"ESP32S3","ty":0,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"14.0.0.1","t":"tasmota_209114","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"bat":0,"dslp":0,"sho":[],"sht":[],"ver":1}
16:44:49.667 RSL: sensors = {"sn":{"Time":"2024-05-21T16:44:49"},"ver":1}
16:44:58.612 WIF: Prüfe Verbindung…
16:45:18.619 WIF: Prüfe Verbindung…
16:45:36.677 -------------------- sende email
16:45:36.678 SCR: performs "sendmail [*:*:*:*:*:to@web.de:TEST] Bodytest"
16:45:36.678 SRC: Rule
16:45:36.679 CMD: Grp 0, Cmd 'SENDMAIL', Idx 1, Len 48, Pld -99, Data '[*:*:*:*:*:to@web.de:TEST] Bodytest'
16:45:55.210 RSL: RESULT = {"Sendmail":"Wrong parameters"}
16:45:57.015 WIF: Prüfe Verbindung…
16:46:06.569 CMD: weblog
16:46:06.570 SRC: WebConsole from 192.168.178.34
16:46:06.571 CMD: Grp 0, Cmd 'WEBLOG', Idx 1, Len 0, Pld -99, Data ''
16:46:06.572 RSL: RESULT = {"WebLog":4}
16:46:17.022 WIF: Prüfe Verbindung…
```
